src/api/app/controllers/webui/package_controller.rb

```diff
@@ -645,83 +642,57 @@ def save_file
     file_url = params[:file_url]
     filename = params[:filename]
 
-    if file.present? # we are getting an uploaded file
-      filename = file.original_filename if filename.blank?
-
-      unless valid_file_name?(filename)
-        if request.xhr?
-          render(json: { error: "'#{filename}' is not a valid filename." }, status: 400)
-        else
-          redirect_to(:back, :error => "'#{filename}' is not a valid filename.")
-        end
-        return
-      end
+    errors = []
 
-      begin
+    begin
+      if file.present?
+        # We are getting an uploaded file
+        filename = file.original_filename if filename.blank?
         @package.save_file(file: file, filename: filename, :comment => params[:comment])
-      rescue Timeout::Error => e
-        if request.xhr?
-          render(json: { error: 'Timeout when saving file. Please try again.'}, status: 400)
-        else
-          redirect_to(:back, :error => 'Timeout when saving file. Please try again.')
+      elsif file_url.present?
+        # we have a remote file URI, so we have to download and save it
+        services = @package.services
+
+        # detects automatically git://, src.rpm formats
+        services.addDownloadURL(file_url, filename)
+
+        unless services.save
+          errors << "Failed to add file from URL '#{file_url}'"
         end
-        return
-      rescue Suse::ValidationError,
-             ActiveXML::Transport::Error => e
-        if request.xhr?
-          render(json: { error: e.summary }, status: 400)
+      else
+        # No file is provided so we just create an empty new file (touch)
+        if filename.present?
+          @package.save_file(filename: filename)
         else
-          redirect_to(:back, :error => e.summary)
+          errors << 'No file or URI given'
         end
-        return
       end
-    # update package timestamp and reindex sources
-    elsif file_url.present? # we have a remote file uri
-      return unless add_file_url(file_url, filename)
-    else
-      return unless add_file_filename(filename)
-    end
-    if request.xhr?
-      render(json: { status: 'ok' })
-    else
-      redirect_to({:action => :show, :project => @project, :package => @package}, :success => "The file #{filename} has been added.")
+
+    rescue APIException, StandardError => exception
```

Comment:
With APIException & StandardError we can catch all possible exceptions (Suse::ValidationError is an APIException and  ActiveXML::Transport::Error is a StandardError etc.)